### DISCUSSION TITLE
Upgrade python versions in monai-deploy MAP Base Image

### DISCUSSION THREAD

**Discussion by AnilMistry** [Project-MONAI/monai-deploy-app-sdk#479] [Q&A]
My segmentation application requires python >=3.9 for certain packages, but I'm limitted to CUDA 11.4. on our deployment platform. I'm tackling with a multi-stage docker build, starting with 'monai-deploy package ...' then running a docker file using that first MAP as a base image. 

I'm using monai-deploy 0.5.1

1. monai-deploy package app (using base image  nvcr.io/nvidia/pytorch:21.07-py3 which has python3.8)
`Monai-deploy package autosegct/app/ -t autosegct/map_init:0.0.1 -l DEBUG -b   nvcr.io/nvidia/pytorch:21.07-py3`

3. docker build -t myapp/map:0.0.1 ./Dockerfile 

```
# Use initial MAP for build
FROM autosegct/map_init:0.0.1 AS build

# Update to python 3.9
ENV PYTHONUNBUFFERED 1
ENV DEBIAN_FRONTEND=noninteractive
ENV TZ=Europe/London

RUN apt-get update &&  apt-get install --no-install-recommends --no-install-suggests -y \
     python3.9 \
     python3-pip \
     && rm -rf /var/lib/apt/lists/*

 # Create symbolic links for Python and pip
 RUN ln -sf /usr/bin/python3.9 /usr/bin/python \
     && ln -sf /usr/bin/python3.9 /usr/bin/python3 \
     && ln -sf /usr/bin/pip3 /usr/bin/pip

RUN echo alias python='/usr/bin/python3.9' >> ~/.bashrc \
    && echo alias python3='/usr/bin/python3.9' >> ~/.bashrc \
    && echo alias pip='/usr/bin/pip3' >> ~/.bashrc \
    && echo alias pip3='/usr/bin/pip3' >> ~/.bashrc

# Add dcm2niix to MAP
WORKDIR /bin
RUN apt-get update &&  apt-get install --no-install-recommends --no-install-suggests -y \
    git build-essential cmake pigz wget


RUN git clone https://github.com/rordenlab/dcm2niix.git --branch master --single-branch \
	&& cd dcm2niix \
	&& mkdir build && cd build \
	&& cmake .. \
	&& make

ENV PATH="$PATH:/bin/dcm2niix/build/bin"

# Install Python packages
WORKDIR /home

RUN apt-get update && apt-get install --no-install-recommends --no-install-suggests -y \
    libsm6 libxrender-dev libxext6 ffmpeg

RUN apt-get install -y libcairo2-dev

# get requirements from command line --build-arg, move to /home/requirements.txt
ARG requirements
ADD $requirements ./

RUN python3.9 -m pip install --upgrade setuptools wheel \
    && python3.9 -m pip install -r ./requirements.txt 

# Workaround for opencv package issue
# see here: https://stackoverflow.com/questions/72706073/attributeerror-partially-initialized-module-cv2-has-no-attribute-gapi-wip-gs
RUN python3.9 -m pip uninstall -y opencv-python
RUN python3.9 -m pip install opencv-python==4.5.5.64
```

This fails to run correctly, reverts to python3.8 again when running app.py, instead of the python3.9 we tried to symlink. 
I believe this is due to the follow part of the MAP build, which enforces python.

https://github.com/Project-MONAI/monai-deploy-app-sdk/blob/978f28c62d00f653beb895176947a631c342c69b/monai/deploy/packager/templates.py#L44

..which I assume builds the monai-exec file which is run by `monai-deploy run` 

Is there a way of resolving this to update a docker base image to python3.9? 
Or are there any base images for MAPs available with python>=3.9 as default but also compatible with CUDA v11.4? 

Many thanks in advance!

**Comment by MMelQin** [comment]
The App SDK is being updated to user Holoscan SDK v1.0.3, published on [pypi](https://pypi.org/project/holoscan/), and as such CUDA 12 runtime is required in App SDK v1.0.0.

For issue with v0.5.1, with the aforementioned templates.py, user's site-packages is inspected and the App SDK modules are copied to the MAP(container image)'s folders. This file can be modified once the App SDK is pip installed for newer versions of Python.

App SDK v1.0.0 will actually pip install the App SDK when building the MAP container image instead.
